main.js

```diff
@@ -9,6 +11,57 @@ let mediaStream
 let thinkingId;
 const rawPCM16WorkerName = "raw-pcm-16-worker";
 
+// Authentication utilities
+
+const fetchServerAccessToken = async () => {
+    const nowSeconds = Math.floor(Date.now() / 1000);
+    const assertionHeader = {
+        alg: "RS256",
+        typ: "JWT",
+    };
+    const payload = {
+        sub: OAUTH_CLIENT_UUID,
+        iss: OAUTH_CLIENT_UUID,
+        aud: `https://${REGION}.api.nabla.com/v1/core/server/oauth/token`,
+        exp: nowSeconds + 60,
+        iat: nowSeconds,
+    };
+    const jwtAssertion = KJUR.jws.JWS.sign(assertionHeader.alg, JSON.stringify(assertionHeader), JSON.stringify(payload), OAUTH_CLIENT_PRIVATE_KEY);
+    const response = await fetch(`https://${REGION}.api.nabla.com/v1/core/server/oauth/token`, {
+        method: 'POST',
+        headers: {
+            'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+            grant_type: "client_credentials",
+            client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+            client_assertion: jwtAssertion,
+        })
+    });
+
+    return response.json();
+};
+
+let serverAccessTokenCache = {
+    accessToken: null,
+    expiresAt: 0,
+};
+
+const getOrRefetchServerAccessToken = async () => {
+    const nowSeconds = Math.floor(Date.now() / 1000);
+
+    if (
+        serverAccessTokenCache.accessToken &&
+        nowSeconds < serverAccessTokenCache.expiresAt
+    ) {
+        return serverAccessTokenCache.accessToken;
+    }
+
+    const { access_token, expires_in } = await fetchServerAccessToken();
+    serverAccessTokenCache.accessToken = access_token;
+    serverAccessTokenCache.expiresAt = nowSeconds + expires_in - 5;
```

Comment:
nit but I'd move the `- 5` to the moment where you read the value (as a leeway for that specific usage) and keep the stored `expiresAt` exactly equal to `nowSeconds + expires_in`